### ISSUE TITLE
ECS: ship components json file abilities parser.

### ISSUE THREAD

**Issue by se5a** [Pulsar4xDevs/Pulsar4x#74]
see https://github.com/Pulsar4xDevs/Pulsar4x/issues/65

Currently (Alpha4) we're using a bunch of lists for the Component Abilities.
It'd be better if we could put some script in the json file (ie allow a line such as ability value = tech guid level * 100)

if we did this we'd have to change how tech works as well, currently we don't have 'levels' for techs, each 'level' is a tech of it's own with its own guid. 

This is part of what we'd need: 
http://en.wikipedia.org/wiki/Shunting-yard_algorithm

**Comment by se5a** [comment]
The parser would have to do two things:

Tell the Component Design UI what to display by reading the Static Data.
and
Validate whatever the Component Design UI created with the Static Data.

so the parser needs to be engine side due to the validation bit, 
what exactly would it send to the UI?
what exactly would the UI create and send back?

**Comment by se5a** [comment]
what sort of syntax are we going to use?

some keywords/functions it's got to recognize as well as those in the above shunting algo would be:
Range(from, too, step)
Guid(xxxx-xxxx-xxxx-xxxx) or see: https://github.com/Pulsar4xDevs/Pulsar4x/issues/73
each of the AbilityTypes in the enum AbilityTypes or would we get away with just having "Ability" as a keyword? maybe we need to wargame a few components and see how they look...

**Comment by se5a** [comment]
I'm glad to report that this is mostly implemented!!!

I found that I was just fighting myself with not having this, and keeped finding myself in positions where I'd have to do some impossible code gymnastics to keep from referencing mod data in the hardcode. 

I stumbled across NCalc which looked to be powerful enough to do the task, yet not too powerful so as to have to worry about sandboxing anything. 
The hardest part was ensuring that where formulaA which depended on other formulaB existed, that when the result for formulaB was changed (ie user changes the size of an engine) that formulaA would be updated. 
There are still some things to do: 
Ensure we cant get into a dependency loop.
More Functions.
Documentation.
Fix broken Tech json. (I added/changed the TechSD)
I'll create new issues for the above and mark this as closed.
